src/metatags/seo-checks.js

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {
+  DESCRIPTION, TITLE, H1, TAG_LENGTHS, ISSUE, ISSUE_DETAILS, SEO_IMPACT, HIGH,
+  SEO_RECOMMENDATION, SHOULD_BE_PRESENT, TITLE_LENGTH_SUGGESTION,
+  DESCRIPTION_LENGTH_SUGGESTION, H1_LENGTH_SUGGESTION, MODERATE,
+  ONE_H1_ON_A_PAGE, UNIQUE_ACROSS_PAGES, DUPLICATES, MULTIPLE_H1_ON_PAGE,
+} from './constants.js';
+
+class SeoChecks {
+  constructor(log) {
+    this.log = log;
+    this.detectedTags = {};
+    this.allTags = {
+      [TITLE]: {},
+      [DESCRIPTION]: {},
+      [H1]: {},
+    };
+  }
+
+  /**
+   * Capitalises the first character of a given string
+   * @param str
+   * @returns {string}
+   */
+  static #capitalizeFirstLetter(str) {
+    return str.charAt(0).toUpperCase() + str.slice(1);
+  }
+
+  /**
+   * Checks for missing tags on the page and adds to detected tags array if found lacking.
+   * @param {string} url - The URL of the page.
+   * @param {object} pageTags - An object containing the tags of the page.
+   */
+  checkForMissingTags(url, pageTags) {
+    [TITLE, DESCRIPTION, H1].forEach((tagName) => {
+      if (pageTags[tagName] === undefined
```

Comment:
should this use `hasText`? what about zero-length strings? this may produce an issue later in the capitalize function. in addition you could make the capitalize function self-guarding.

Reply:
Here we are checking for missing tags, the identifier for which is undefined tag value. If pageTags[tagName] is an empty string then it would be detected in emptyTag check in `checkForTagsLength` method.
The capitalize function is always called with tagName (_not pageTags[tagName]_) which will always be title, description or h1.
As a safety guard, I will add a hasText check to the capitalizeFirstLetter method.